src/control/lib/netdetect/netdetect.go

```diff
@@ -0,0 +1,163 @@
+// specify no topology file and hwloc will perform queries against the actual
+// local system.
+func initLib(xmlTopology string) (C.hwloc_topology_t, error) {
+	var topology C.hwloc_topology_t
+
+	status := C.hwloc_topology_init(&topology)
+	if status != 0 {
+		return nil, errors.New("hwloc_topology_init failure")
+	}
+
+	if (xmlTopology != "") {
+		status := C.hwloc_topology_set_xml(topology, C.CString(xmlTopology))
+		if (status != 0) {
+			return nil, errors.New("hwloc_topology_set_xml failure")
+		}
+	}
+
+	status = C.hwloc_topology_set_flags(topology, C.HWLOC_TOPOLOGY_FLAG_IO_DEVICES)
+	if status != 0 {
+		return nil, errors.New("hwloc_topology_set_flags failure")
+	}
+
+	status = C.hwloc_topology_load(topology)
+	if status != 0 {
+		return nil, errors.New("hwloc_topology_load failure")
+	}
+	return topology, nil
+}
+
+// cleanUp closes out the hwloc resources
+func cleanUp(topology C.hwloc_topology_t) {
+	if (topology != nil) {
+		C.hwloc_topology_destroy(topology)
+	}
+}
+
+// GetAffinityForNetworkDevices searches the system topology for each network
+// device specified by the input string and returns corresponding cpuset and
+// nodeset bitmaps as strings.  When converted back to hwloc_bitmap_t via
+// hwloc_bitmap_sscanf() these bitmaps are used by the hwloc API to bind a
+// thread to processing units that are closest to the given network device.
+//
+// The input netNames string has the form "device0,device1,...deviceN"
+// where 'device' specifies a network device such as "eth0", "eth1", or "ib0"
+// The return string has the form:
+// "dev0:cpuset:nodeset;dev1:cpuset:nodeset; ... devN:cpuset:nodeset"
+// where dev is the name of the network device and cpuset and nodeset are
+// the corresponding bitmap strings.
+//
+// Network device names that are not found in the topology are ignored.
+// The order of network devices in the return string depends on the natural
+// order in the system topology and does not depend on the order specified
+// by the input string.
+func GetAffinityForNetworkDevices(netNames string,
+	xmlTopology string) (string, error) {
+
+	var affinity string
+	var node C.hwloc_obj_t
+	var i C.uint
+	var cpuset *C.char
+	var nodeset *C.char
+
+	topology, err := initLib(xmlTopology)
+	if (err != nil) {
+		log.Debugf("Error from initLib %v", err)
+		return affinity,
+			errors.New("unable to initialize hwloc library")
+	}
+
+	depth := C.hwloc_get_type_depth(topology, C.HWLOC_OBJ_OS_DEVICE)
+	if depth != C.HWLOC_TYPE_DEPTH_OS_DEVICE {
+		return affinity,
+			errors.New("hwloc_get_type_depth returned invalid value")
+	}
+
+	numObj := C.hwloc_get_nbobjs_by_depth(topology, C.uint(depth))
+	// for any OS object found in the network device list,
+	// detect and store the cpuset and nodeset of the ancestor node
+	// containing this object
+	for i = 0; i < numObj; i++ {
+		node = C.hwloc_get_obj_by_depth(topology, C.uint(depth), i)
+		if node != nil && strings.Contains(netNames, C.GoString(node.name)) {
```

Comment:
I think this may be a potential source of bugs. What happens when you have a `node.name` of `eth10` but your `netNames` is `eth100,eth3`?

It would be more reliable to create a map of netNames and check the node.name against that:
```go
        netNames := map[string]struct{}
        for _, deviceName := range deviceNames {
                netNames[deviceName] = struct{}{} // creates an empty struct; doesn't use any space
        }
...
        for i = 0; i < numObj; i++ {
                if node == nil {
                        continue
                }
                if _, found := netNames[C.GoString(node.name)]; !found {
                        continue
                }
....
```

Reply:
I went with simple at the time, and I agree with you that it is potentially unsafe.  I'll use the approach you used here since that won't be potentially fragile.